### PR TITLE
Remove randomness from port assignment during coverage collection.

### PR DESCRIPTION
**Pull request by Hixie**
Also, defer to test package for throttling (this may require a test
package update as well).

Also, add a lot more instrumentation to --verbose mode for tests, and
fix minor trivial things here and there, and add error handling in
more places.

Also, refactor how coverage works to be simpler and not use statics.

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 1]
```diff
@@ -60,36 +23,51 @@ class CoverageCollector {
...
-    return future;
+  /// This should be called when the code whose coverage data is being collected
+  /// has been run to completion so that all coverage data has been recorded.
+  ///
+  /// The returned [Future] completes when the coverage is collected.
+  Future<Null> collectCoverage(Process process, InternetAddress host, int port) {
+    assert(process != null);
+    assert(port != null);
+
+    Completer<Null> completer = new Completer<Null>();
```
Comment (abarth) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: Maybe just make the function `async`

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -154,81 +197,135 @@ class FlutterPlatform extends PlatformPlugin {
...
         case _InitialResult.connected:
+          printTrace('test $ourTestCount: process with pid ${process.pid} connected to test harness');
           WebSocket testSocket = await webSocket.future;
 
           Completer<Null> harnessDone = new Completer<Null>();
           StreamSubscription<dynamic> harnessToTest = controller.stream.listen(
-            (dynamic event) { testSocket.add(JSON.encode(event)); },
+            (dynamic event) {
+              final String encodedEvent = JSON.encode(event);
+              printTrace('test $ourTestCount: harness => ${process.pid}: ${_truncate(encodedEvent, 80, "...")}');
```
Comment (abarth) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: I'd skip the traces on each protocol message.

**Review by abarth — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 3]
```diff
@@ -286,14 +378,11 @@ void main() {
   }
 
 
-  Future<Process> _startProcess(String executable, String testPath, { String packages, int observatoryPort }) {
+  Future<Process> _startProcess(String executable, String testPath, { String packages, bool enableObservatory: false }) {
     assert(executable != null); // Please provide the path to the shell in the SKY_SHELL environment variable.
     List<String> arguments = <String>[];
-    if (observatoryPort != null) {
-      arguments.add('--observatory-port=$observatoryPort');
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: We need a way to specify the observatory port, for use inside Google.

**Comment by Hixie** [comment]
A regression from this got fixed in https://github.com/flutter/flutter/pull/7558

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 4]
```diff
@@ -99,52 +143,51 @@ class FlutterPlatform extends PlatformPlugin {
...
+          printTrace('test $ourTestCount: ensuring end-of-process for shell');
           process.kill();
-        int exitCode = await process.exitCode;
-        subprocessActive = false;
-        if (!controllerSinkClosed && exitCode != 0) {
-          String message = _getErrorMessage(_getExitCodeMessage(exitCode, 'after tests finished'), testPath, shellPath);
-          controller.sink.addError(message);
+          final int exitCode = await process.exitCode;
+          subprocessActive = false;
+          if (!controllerSinkClosed && exitCode != -15) {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: `-15`?

Comment (Hixie) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: https://github.com/flutter/flutter/pull/7571

**Pull request merged**